### PR TITLE
[Refactor] New API (work in progress)

### PR DESCRIPTION
**Pull request by fchollet**
Do not merge yet, this branch should replace master once it's ready. Work in progress for now.

### CODE REVIEW

**Comment by fchollet** [comment]
The refactor has three broad goals:

1) Implement a revised user-facing API.
2) Upgrade the assumptions in the codebase to follow TF 2.0 APIs. The previous codebase largely came from multi-backend Keras and thus makes assumptions that are not always applicable in TF 2.0 (e.g. no support for subclassed models (which are non-serializable), no inspection of `keras.metrics.Metric` instances, old metric naming conventions...)
3) Simplify the codebase by reducing the number of internal classes and the internal API surface, and generally ironing out communication patterns across internal classes.
